Update pydicom to version 2.2

Updating causes breaking changes.

@taleinat thank you for reaching out! 
Updating `pydicom` past 2.1.2 causes the tests to fail due to various changes made to the values returned for Siemens DWI series, as well as `PersonName` (`PN`) data elements:
```bash
====================================================================================== short test summary info ======================================================================================
FAILED tests/test_header.py::HeaderTestCase::test_get_data_elements_with_exclude_list - AssertionError: 83 != 59
FAILED tests/test_header.py::HeaderTestCase::test_get_data_elements_with_vr_list - AssertionError: 26 != 20
FAILED tests/test_image.py::ImageTestCase::test_b_matrix - AssertionError: False is not true
FAILED tests/test_image.py::ImageTestCase::test_b_vector - ValueError: shapes (6,) and (3,3) not aligned: 6 (dim 0) != 3 (dim 0)
FAILED tests/test_image.py::ImageTestCase::test_q_vector - ValueError: shapes (6,) and (3,3) not aligned: 6 (dim 0) != 3 (dim 0)
FAILED tests/test_image.py::ImageTestCase::test_voxel_space_b_matrix - ValueError: shapes (6,) and (3,3) not aligned: 6 (dim 0) != 3 (dim 0)
FAILED tests/data_elements/test_person_name.py::PersonNameTestCase::test_parse_value - AssertionError: {'name_prefix': '', 'given_name': '', 'mid[47 chars]: ''} != {}
FAILED tests/data_elements/test_private_data_element.py::SiemensTestCase::test_parse_value - AssertionError: (1.0, 1.0, 1.0, 1.0, 1.0, 1.0) != array([[1., 1., 1.],
==================================================================== 8 failed, 269 passed, 212 skipped, 23036 warnings in 8.12s =====================================================================
```
It seems `pydicom` added some automatic parsing functionality that behaves differently than `dicom_parser`. 
What's required now is to:
- [ ] Check the differences between how `pydicom` parses out the relevant elements before 2.2.* and after.
- [ ] Understand why it is different than `dicom_parser`'s logic and choose the correct/preferred parsing strategy.
- [ ] Change `dicom_parser`'s parsing logic accordingly.
- [ ] Update requirements to require `dicom_parser>=2.2.*` or customize parsing logic based on `pydicom` version and remove the existing limitation in the dependencies.

Hope this is clearer, any help would be greatly appreciated.

--- a/setup.cfg
+++ b/setup.cfg
@@ -36,7 +36,7 @@ python_requires = >=3.6, <4
 install_requires =
     numpy;python_version=="3.6"
     numpy~=1.21;python_version>"3.6"
-    pydicom==2.1.2
+    pydicom>=2.1,<2.3
 
 [options.extras_require]
 magic =

--- a/setup.cfg
+++ b/setup.cfg
@@ -36,7 +36,7 @@ python_requires = >=3.6, <4
 install_requires =
     numpy;python_version=="3.6"
     numpy~=1.21;python_version>"3.6"
-    pydicom>=2.1,<2.3
+    pydicom>=2.1,<2.4
 
 [options.extras_require]
 magic =

--- a/src/dicom_parser/data_elements/person_name.py
+++ b/src/dicom_parser/data_elements/person_name.py
@@ -35,7 +35,7 @@ class PersonName(DataElement):
             Parsed person name components
         """
         if not value:
-            value = PydicomPersonName('')
+            value = PydicomPersonName("")
         elif not isinstance(value, PydicomPersonName):
             return value
         return {

--- a/src/dicom_parser/data_elements/person_name.py
+++ b/src/dicom_parser/data_elements/person_name.py
@@ -34,10 +34,11 @@ class PersonName(DataElement):
         dict
             Parsed person name components
         """
-
-        if isinstance(value, PydicomPersonName):
-            return {
-                component: getattr(value, component)
-                for component in self.COMPONENTS
-            }
-        return value if value else {}
+        if not value:
+            value = PydicomPersonName('')
+        elif not isinstance(value, PydicomPersonName):
+            return value
+        return {
+            component: getattr(value, component)
+            for component in self.COMPONENTS
+        }

--- a/src/dicom_parser/data_elements/private_data_element.py
+++ b/src/dicom_parser/data_elements/private_data_element.py
@@ -8,6 +8,7 @@ from typing import Any
 from dicom_parser.data_element import DataElement
 from dicom_parser.utils.siemens.private_tags import (
     parse_siemens_b_matrix,
+    parse_siemens_b_matrix_multiple,
     parse_siemens_bandwith_per_pixel_phase_encode,
     parse_siemens_csa_header,
     parse_siemens_gradient_direction,
@@ -29,6 +30,9 @@ TAG_TO_PARSER = {
     ("0019", "1029"): parse_siemens_slice_timing,
     ("0029", "1010"): parse_siemens_csa_header,
     ("0029", "1020"): parse_siemens_csa_header,
+}
+TAG_TO_MULTIVALUE_PARSER = {
+    ("0019", "1027"): parse_siemens_b_matrix_multiple,
 }
 
 
@@ -76,3 +80,19 @@ class PrivateDataElement(DataElement):
 
         # Otherwise, simply return the raw value.
         return value
+
+    def parse_values(self) -> Any:
+        """
+        Tries to parse private data element values using a custom function.
+
+        Returns
+        -------
+        Any
+            This instance's parsed value or values
+        """
+        if self.value_multiplicity > 1:
+            # Try to call a custom parser function.
+            method: FunctionType = TAG_TO_MULTIVALUE_PARSER.get(self.tag)
+            if method:
+                return method(self.raw.value)
+        return super().parse_values()

--- a/src/dicom_parser/utils/siemens/private_tags.py
+++ b/src/dicom_parser/utils/siemens/private_tags.py
@@ -3,7 +3,7 @@ Siemens specific private tags they may not be accessible by keyword using
 `pydicom <https://github.com/pydicom/pydicom>`_.
 """
 import array
-from typing import Tuple, Union
+from typing import List, Tuple, Union
 
 import numpy as np
 from dicom_parser.utils.siemens import messages
@@ -134,6 +134,7 @@ def parse_siemens_b_matrix(value: Union[float, bytes]) -> np.ndarray:
     if isinstance(value, float):
         return value
     elif isinstance(value, bytes):
+        # pydicom < 2.2 returns the VR "UN" and the value as an array of bytes.
         raw = list(array.array("d", value))
         return np.array(raw)[B_MATRIX_INDICES].reshape(3, 3)
     else:
@@ -141,6 +142,24 @@ def parse_siemens_b_matrix(value: Union[float, bytes]) -> np.ndarray:
             name="B_matrix", valid_types=(bytes, float), value=value
         )
         raise ValueError(message)
+
+
+def parse_siemens_b_matrix_multiple(value: List[float]):
+    """
+    Parses the Siemens B matrix header field value.
+
+    Parameters
+    ----------
+    value : List[float]
+        Raw B matrix header field value
+
+    Returns
+    -------
+    np.ndarray
+        Parsed B matrix
+    """
+    # pydicom >= 2.2 returns the VR "FD" and the value as a list of floats.
+    return np.array(value)[B_MATRIX_INDICES].reshape(3, 3)
 
 
 def b_matrix_to_q_vector(

--- a/tests/data_elements/fixtures.py
+++ b/tests/data_elements/fixtures.py
@@ -56,7 +56,13 @@ LONG_STRINGS = {
 
 #: Expected parsed values to test person names (PN) data elements with.
 PERSON_NAMES = {
-    "ReferringPhysicianName": {},
+    "ReferringPhysicianName": {
+        "name_prefix": "",
+        "given_name": "",
+        "middle_name": "",
+        "family_name": "",
+        "name_suffix": "",
+    },
     "OperatorsName": {
         "name_prefix": "",
         "given_name": "",

--- a/tests/test_data_element.py
+++ b/tests/test_data_element.py
@@ -55,16 +55,16 @@ class DataElementTestCase(TestCase):
                 value = self.TEST_CLASS(raw).value
                 if isinstance(value, np.ndarray):
                     self.assertTrue(np.array_equal(value, expected))
-                elif (
-                    self._is_nonempty_float_sequence(expected) and
-                    self._is_nonempty_float_sequence(value)
-                ):
+                elif self._is_nonempty_float_sequence(
+                    expected
+                ) and self._is_nonempty_float_sequence(value):
                     self.assertEqual(len(value), len(expected))
                     self.assertEqual(type(value), type(expected))
-                    self.assertTrue(all(
-                        math.isclose(a, b)
-                        for a, b in zip(value, expected)
-                    ))
+                    self.assertTrue(
+                        all(
+                            math.isclose(a, b) for a, b in zip(value, expected)
+                        )
+                    )
                 else:
                     self.assertEqual(value, expected)
 
@@ -89,15 +89,14 @@ class DataElementTestCase(TestCase):
     @classmethod
     def _is_nonempty_float_sequence(cls, value):
         return (
-            isinstance(value, Sequence) and
-            value and
-            isinstance(value[0], float)
+            isinstance(value, Sequence)
+            and value
+            and isinstance(value[0], float)
         )
 
     def assertEqualFloatSequences(self, first, second, msg=None):
         self.assertEqual(len(first), len(second), msg=msg)
         self.assertEqual(type(first), type(second), msg=msg)
-        self.assertTrue(all(
-            math.isclose(a, b)
-            for a, b in zip(first, second)
-        ), msg=msg)
+        self.assertTrue(
+            all(math.isclose(a, b) for a, b in zip(first, second)), msg=msg
+        )

--- a/tests/test_data_element.py
+++ b/tests/test_data_element.py
@@ -55,16 +55,11 @@ class DataElementTestCase(TestCase):
                 value = self.TEST_CLASS(raw).value
                 if isinstance(value, np.ndarray):
                     self.assertTrue(np.array_equal(value, expected))
-                elif self._is_nonempty_float_sequence(
-                    expected
-                ) and self._is_nonempty_float_sequence(value):
-                    self.assertEqual(len(value), len(expected))
-                    self.assertEqual(type(value), type(expected))
-                    self.assertTrue(
-                        all(
-                            math.isclose(a, b) for a, b in zip(value, expected)
-                        )
-                    )
+                elif all([
+                    self._is_nonempty_float_sequence(value),
+                    self._is_nonempty_float_sequence(expected),
+                ]):
+                    self.assertEqualFloatSequences(value, expected)
                 else:
                     self.assertEqual(value, expected)
 

--- a/tests/test_data_element.py
+++ b/tests/test_data_element.py
@@ -1,7 +1,8 @@
 """
 Definition of the :class:`DataElementTestCase` class.
 """
-from typing import Tuple, Union
+import math
+from typing import Tuple, Union, Sequence
 from unittest import TestCase
 
 import numpy as np
@@ -54,6 +55,16 @@ class DataElementTestCase(TestCase):
                 value = self.TEST_CLASS(raw).value
                 if isinstance(value, np.ndarray):
                     self.assertTrue(np.array_equal(value, expected))
+                elif (
+                    self._is_nonempty_float_sequence(expected) and
+                    self._is_nonempty_float_sequence(value)
+                ):
+                    self.assertEqual(len(value), len(expected))
+                    self.assertEqual(type(value), type(expected))
+                    self.assertTrue(all(
+                        math.isclose(a, b)
+                        for a, b in zip(value, expected)
+                    ))
                 else:
                     self.assertEqual(value, expected)
 
@@ -74,3 +85,19 @@ class DataElementTestCase(TestCase):
             self.skipTest(self.SKIP_MESSAGE)
         element = self.TEST_CLASS(self.raw_element)
         self.assertFalse(element.is_private)
+
+    @classmethod
+    def _is_nonempty_float_sequence(cls, value):
+        return (
+            isinstance(value, Sequence) and
+            value and
+            isinstance(value[0], float)
+        )
+
+    def assertEqualFloatSequences(self, first, second, msg=None):
+        self.assertEqual(len(first), len(second), msg=msg)
+        self.assertEqual(type(first), type(second), msg=msg)
+        self.assertTrue(all(
+            math.isclose(a, b)
+            for a, b in zip(first, second)
+        ), msg=msg)

--- a/tests/test_data_element.py
+++ b/tests/test_data_element.py
@@ -49,12 +49,13 @@ class DataElementTestCase(TestCase):
             self.skipTest(self.SKIP_MESSAGE)
         values = self.get_values()
         for key, expected in values.items():
-            raw = self.get_raw_element(key)
-            value = self.TEST_CLASS(raw).value
-            if isinstance(value, np.ndarray):
-                self.assertTrue(np.array_equal(value, expected))
-            else:
-                self.assertEqual(value, expected)
+            with self.subTest(key=key):
+                raw = self.get_raw_element(key)
+                value = self.TEST_CLASS(raw).value
+                if isinstance(value, np.ndarray):
+                    self.assertTrue(np.array_equal(value, expected))
+                else:
+                    self.assertEqual(value, expected)
 
     def test_repr(self):
         if not self.SAMPLE_KEY:

--- a/tests/test_header.py
+++ b/tests/test_header.py
@@ -84,6 +84,9 @@ class HeaderTestCase(TestCase):
     #: `keyword_contains()` call.
     KEYWORD_CONTAINS = {"time": 10, "DATE": 7, "abcdef": 0}
 
+    # Version info for pydicom, converted to integers.
+    PYDICOM_VERSION_INFO = tuple(map(int, pydicom.__version_info__))
+
     def setUp(self):
         self.raw = pydicom.dcmread(TEST_IMAGE_PATH)
         self.header = Header(self.raw)
@@ -271,7 +274,8 @@ class HeaderTestCase(TestCase):
         vr = [ValueRepresentation.IS, ValueRepresentation.CS]
         result = self.header.get_data_elements(value_representation=vr)
         self.assertIsInstance(result, list)
-        self.assertEqual(len(result), 20)
+        expected = 20 if self.PYDICOM_VERSION_INFO < (2, 2) else 26
+        self.assertEqual(len(result), expected)
 
     def test_get_data_elements_with_vr_tuple(self):
         vr = ValueRepresentation.DA, ValueRepresentation.TM
@@ -288,7 +292,8 @@ class HeaderTestCase(TestCase):
         ]
         result = self.header.get_data_elements(exclude=exclude)
         self.assertIsInstance(result, list)
-        self.assertEqual(len(result), 59)
+        expected = 59 if self.PYDICOM_VERSION_INFO < (2, 2) else 83
+        self.assertEqual(len(result), expected)
 
     def test_get_data_elements_with_exclude_tuple(self):
         exclude = ValueRepresentation.DA, ValueRepresentation.TM

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 [tox]
 envlist =
-    py{36,37,38,39,310}{,-nomagic,-pandas}
+    py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}
     lint
     cov-report
     cov-clean
@@ -10,21 +10,24 @@ isolated_build = True
 
 [gh-actions]
 python =
-    3.6: py36, py36-nomagic, py36-pandas
-    3.7: py37, py37-nomagic, py37-pandas
-    3.8: py38, py38-nomagic, py38-pands, lint, cov-report, cov-clean
-    3.9: py39, py39-nomagic, py39-pandas
-    3.10: py310, py310-nomagic, py310-pandas
+    3.6: py36-pydicom21, py36-pydicom22, py36-pydicom21-nomagic, py36-pydicom22-nomagic, py36-pydicom21-pandas, py36-pydicom22-pandas
+    3.7: py37-pydicom21, py37-pydicom22, py37-pydicom21-nomagic, py37-pydicom22-nomagic, py37-pydicom21-pandas, py37-pydicom22-pandas
+    3.8: py38-pydicom21, py38-pydicom22, py38-pydicom21-nomagic, py38-pydicom22-nomagic, py38-pydicom21-pandas, py38-pydicom22-pandas, lint, cov-report, cov-clean
+    3.9: py39-pydicom21, py39-pydicom22, py39-pydicom21-nomagic, py39-pydicom22-nomagic, py39-pydicom21-pandas, py39-pydicom22-pandas
+    3.10: py310-pydicom21, py310-pydicom22, py310-pydicom21-nomagic, py310-pydicom22-nomagic, py310-pydicom21-pandas, py310-pydicom22-pandas
 
 [testenv]
-deps = -e .[test]
+deps =
+    -e .[test]
+    pydicom21: pydicom>=2.1,<2.2
+    pydicom22: pydicom>=2.2,<2.3
 depends =
-    py{36,37,38,39,310}{,-nomagic,-pandas}: cov-clean, lint
+    py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}: cov-clean, lint
 setenv =
-    py{36,37,38,39,310}{,-nomagic,-pandas}: COVERAGE_FILE=.coverage.{envname}
+    py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}: COVERAGE_FILE=.coverage.{envname}
 commands_pre =
-    py{36,37,38,39,310}-nomagic: pip uninstall -y python-magic
-    py{36,37,38,39,310}-pandas: pip install -e .[pandas]
+    py{36,37,38,39,310}-pydicom{21,22}-nomagic: pip uninstall -y python-magic
+    py{36,37,38,39,310}-pydicom{21,22}-pandas: pip install -e .[pandas]
 commands = coverage run
 
 [testenv:lint]
@@ -36,7 +39,7 @@ deps = coverage[toml]
 skip_install = true
 parallel_show_output = true
 depends =
-    cov-clean, py{36,37,38,39,310}{,-nomagic,-pandas}
+    cov-clean, py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}
 commands =
     coverage combine
     coverage report

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 [tox]
 envlist =
-    py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}
+    py{36,37,38,39,310}-pydicom{21,22,23}{,-nomagic,-pandas}
     lint
     cov-report
     cov-clean
@@ -10,24 +10,25 @@ isolated_build = True
 
 [gh-actions]
 python =
-    3.6: py36-pydicom21, py36-pydicom22, py36-pydicom21-nomagic, py36-pydicom22-nomagic, py36-pydicom21-pandas, py36-pydicom22-pandas
-    3.7: py37-pydicom21, py37-pydicom22, py37-pydicom21-nomagic, py37-pydicom22-nomagic, py37-pydicom21-pandas, py37-pydicom22-pandas
-    3.8: py38-pydicom21, py38-pydicom22, py38-pydicom21-nomagic, py38-pydicom22-nomagic, py38-pydicom21-pandas, py38-pydicom22-pandas, lint, cov-report, cov-clean
-    3.9: py39-pydicom21, py39-pydicom22, py39-pydicom21-nomagic, py39-pydicom22-nomagic, py39-pydicom21-pandas, py39-pydicom22-pandas
-    3.10: py310-pydicom21, py310-pydicom22, py310-pydicom21-nomagic, py310-pydicom22-nomagic, py310-pydicom21-pandas, py310-pydicom22-pandas
+    3.6: py36-pydicom21, py36-pydicom22, py36-pydicom23, py36-pydicom21-nomagic, py36-pydicom22-nomagic, py36-pydicom23-nomagic, py36-pydicom21-pandas, py36-pydicom22-pandas, py36-pydicom23-pandas
+    3.7: py37-pydicom21, py37-pydicom22, py37-pydicom23, py37-pydicom21-nomagic, py37-pydicom22-nomagic, py37-pydicom23-nomagic, py37-pydicom21-pandas, py37-pydicom22-pandas, py37-pydicom23-pandas
+    3.8: py38-pydicom21, py38-pydicom22, py38-pydicom23, py38-pydicom21-nomagic, py38-pydicom22-nomagic, py38-pydicom23-nomagic, py38-pydicom21-pandas, py38-pydicom22-pandas, py38-pydicom23-pandas, lint, cov-report, cov-clean
+    3.9: py39-pydicom21, py39-pydicom22, py39-pydicom23, py39-pydicom21-nomagic, py39-pydicom22-nomagic, py39-pydicom23-nomagic, py39-pydicom21-pandas, py39-pydicom22-pandas, py39-pydicom23-pandas
+    3.10: py310-pydicom21, py310-pydicom22, py310-pydicom23, py310-pydicom21-nomagic, py310-pydicom22-nomagic, py310-pydicom23-nomagic, py310-pydicom21-pandas, py310-pydicom22-pandas, py310-pydicom23-pandas
 
 [testenv]
 deps =
     -e .[test]
     pydicom21: pydicom>=2.1,<2.2
     pydicom22: pydicom>=2.2,<2.3
+    pydicom23: pydicom>=2.3,<2.4
 depends =
-    py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}: cov-clean, lint
+    py{36,37,38,39,310}-pydicom{21,22,23}{,-nomagic,-pandas}: cov-clean, lint
 setenv =
-    py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}: COVERAGE_FILE=.coverage.{envname}
+    py{36,37,38,39,310}-pydicom{21,22,23}{,-nomagic,-pandas}: COVERAGE_FILE=.coverage.{envname}
 commands_pre =
-    py{36,37,38,39,310}-pydicom{21,22}-nomagic: pip uninstall -y python-magic
-    py{36,37,38,39,310}-pydicom{21,22}-pandas: pip install -e .[pandas]
+    py{36,37,38,39,310}-pydicom{21,22,23}-nomagic: pip uninstall -y python-magic
+    py{36,37,38,39,310}-pydicom{21,22,23}-pandas: pip install -e .[pandas]
 commands = coverage run
 
 [testenv:lint]
@@ -39,7 +40,7 @@ deps = coverage[toml]
 skip_install = true
 parallel_show_output = true
 depends =
-    cov-clean, py{36,37,38,39,310}-pydicom{21,22}{,-nomagic,-pandas}
+    cov-clean, py{36,37,38,39,310}-pydicom{21,22,23}{,-nomagic,-pandas}
 commands =
     coverage combine
     coverage report

